### DISCUSSION TITLE
Turn Visualmode into Multicaret-Insertmode

### DISCUSSION THREAD

**Discussion by comod** [JetBrains/ideavim#438] [Q&A]
Lets say i have a big list and i want to add something at the beginning each selected line and i dont want to use macros it would be great to use multicarets for that.

**Comment by lippfi** [comment]
Caret block selection helps to achieve this. Lets say you have a list of names like:
```{your caret is here}Mike
Kate
John
Bob
....
```
Then you can type `<C-v>GIHello, <Esc>` to prepend "Hello, " before each name in the list

For more complicated cases you can enable multiple cursors in IdeaVim by adding `Plug 'terryma/vim-multiple-cursors'` to your `.ideavimrc` (https://github.com/JetBrains/ideavim/wiki/Emulated-plugins#multiple-cursors)

**Comment by comod** [comment]
`<C-v>GIHello, <Esc>`
unfortunately this does not work (neither in phpstorm nor in nvim; idk why; is it possibe that any vimrc setting block its functionality?)

`terryma/vim-multiple-cursors` also doesnt work as described - i am not sure what i am doing wrong :-/

**Reply by AlexPl292** [reply]
Have you tried to enter the full command? After pressing the capital `I`, vim shoes only a single caret, what may be confusing, however, after pressing `esc` the prefix will be inserted on every line.
If it still doesn't work, could you please share your `~/.ideavimrc` and `.vimrc` files?

**Reply by comod** [reply]
Yes, i tried it with `I` and `esc`

~/.nvimrc
```
nnoremap <SPACE> <Nop>
set number relativenumber
let mapleader=" "
imap jj <Esc>
filetype plugin on
syntax on
```

~/.ideavimrc:
```
source ~/.nvimrc
set which-key
set notimeout
set ideajoin
clipboard ideaput
set ideamarks
set surround
set repeat
set ReplaceWithRegister
set commentary
set exchange
set vim-multiple-cursors
map <F4> o<ESC>p
nnoremap <C-t> :NERDTreeToggle<CR>
map gu <Action>(FindUsages)
map gz <Action>(ShowUsages)
map gp <Action>(FindInPath)
map gf <Action>(Find)
map ä <Action>(ShiftDownAction)
map - <Action>(EditorLineEnd)
```

**Reply by comod** [reply]
`set multiple-cursors`
seems to work, but i am still confused how to use this plugin, especially when i have manually selected text and i simply want to add multicaret of each line - it always puts than one more line at the end on it. and if i want to insert text i need to hit `i` 3 times where can i find a list of actions? are these ones the complete list of actions?:

```
let g:multi_cursor_use_default_mapping=0
let g:multi_cursor_start_word_key      = '<C-n>'
let g:multi_cursor_select_all_word_key = '<A-n>'
let g:multi_cursor_start_key           = 'g<C-n>'
let g:multi_cursor_select_all_key      = 'g<A-n>'
let g:multi_cursor_next_key            = '<C-n>'
let g:multi_cursor_prev_key            = '<C-p>'
let g:multi_cursor_skip_key            = '<C-x>'
let g:multi_cursor_quit_key            = '<Esc>'
```

**Reply by citizenmatt** [reply]
The `multiple-cursors` plugin works in a manner very similar to [terryma/vim-multiple-cursors](https://github.com/terryma/vim-multiple-cursors), although there are some significant differences - it doesn't have any options, and by default, it's got a different set of shortcuts (I don't know why - see [VIM-2178](https://youtrack.jetbrains.com/issue/VIM-2178) for more on the shortcuts).

There are a couple of ways of working. You can repeatedly use the select next/prev shortcut to select the next/previous occurrence of the word under the cursor, using the skip shortcut to skip an occurrence. Or you can select some text first, then use next/prev/skip to select multiple occurrences of matching text. This gives you multiple cursors, with multiple selections, so you're effectively in VISUAL mode, with multiple cursors/selections. You can now use commands like `x` to delete, `c` to change, `r` to replace, `~` to change case, etc. You should be able to perform any command that you'd normally perform in VISUAL mode with a single selection, but applied to all selections. Note that you can't use `i` to insert - that doesn't work in normal VISUAL mode. You can hit `<Esc>` to exit this VISUAL-like mode while still keeping multiple cursors. At this point, you can use `i` to insert. The examples in the [terryma/vim-multiple-cursors](https://github.com/terryma/vim-multiple-cursors) readme demonstrate this.

Alternatively, you can select a block of text, e.g. `vip` to select a paragraph, then use the "next occurrence" shortcut to convert that selection into multiple cursors, with the cursor at the beginning of each line. Again, this is demonstrated at [terryma/vim-multiple-cursors](https://github.com/terryma/vim-multiple-cursors).

It is confusing that the default shortcut set is different to [terryma/vim-multiple-cursors](https://github.com/terryma/vim-multiple-cursors). I don't know where this set of shortcuts came from. I have them remapped in my `~/.ideavimrc`, which you can see in the [VIM-2178](https://youtrack.jetbrains.com/issue/VIM-2178) issue.

**Reply by comod** [reply]
We are closer :-), but it still does not behave as you describe. eg: `vip` + `A-n` + `x` does nothing - it needs one more `x` to start deleting. same with the others; uppercase + `I` works immediately, though. 
i saw this kind of remaps:
`nmap <C-n> <Plug>NextWholeOccurrence`
Where can i find a list of these "Plug"-Action-like-Namings? Especially for "start" (`A-n`). I would like to create a own binding which results in insert mode directly.
